cumulus-client/src/main/java/mil/army/usace/hec/cumulus/client/controllers/CumulusDssFileController.java

```diff
@@ -0,0 +1,174 @@
+            } catch (IOException ex) {
+                throw new CompletionException(ex);
+            }
+        });
+    }
+
+    /**
+     * Generate DSS File.
+     *
+     * @param apiConnectionInfo - connection info
+     * @param downloadRequest   - Download Request object containing start, end, watershed ID, and product IDs
+     * @return Download         - object that can be queried for status updates
+     */
+    public CompletableFuture<Download> generateDssFile(ApiConnectionInfo apiConnectionInfo, DownloadRequest downloadRequest,
+                                                       OAuth2Token token, CumulusDssFileGenerationListener listener) throws CompletionException {
+        return CompletableFuture.supplyAsync(() -> {
+            try {
+                if (downloadRequest == null) {
+                    throw new IOException("Missing required Download Request in order to generate DSS File");
+                }
+                Download initialDownloadData = executeInitialDownloadRequest(apiConnectionInfo, downloadRequest, token);
+                if (listener != null) {
+                    listener.downloadStatusUpdated(initialDownloadData);
+                }
+                return initialDownloadData;
+            } catch (IOException e) {
+                throw new CompletionException(e);
+            }
+        });
+    }
+
+    Download executeInitialDownloadRequest(ApiConnectionInfo apiConnectionInfo, DownloadRequest downloadRequest, OAuth2Token token) throws IOException {
+        String jsonBody = CumulusObjectMapper.mapObjectToJson(downloadRequest);
+        HttpRequestExecutor executor =
+            new CumulusHttpRequestBuilderImpl(apiConnectionInfo, DOWNLOADS_ENDPOINT, token)
+                .enableHttp2()
+                .post()
+                .withBody(jsonBody)
+                .withMediaType(ACCEPT_HEADER_V1);
+        try (HttpRequestResponse response = executor.execute()) {
+            return CumulusObjectMapper.mapJsonToObject(response.getBody(), Download.class);
+        }
+    }
+
+    /**
+     * Download File to specified local file location.
+     *
+     * @param generatedDssFileDownloadData - Download object containing URL of file to download
+     * @param pathToLocalFile              - path in which file will be downloaded to
+     * @param listener                     - CumulusDssFileDownloadListener implementation that listens for download updates
+     * @return CompletableFuture           - Reference to async download process to be handled by caller
+     */
+    public CompletableFuture<Void> downloadToLocalFile(Download generatedDssFileDownloadData, Path pathToLocalFile,
+                                                       CumulusDssFileDownloadListener listener) throws CompletionException {
+        if (generatedDssFileDownloadData == null) {
+            throw new CompletionException(new IOException("Missing download Data"));
```

Comment:
Should not be throwing a CompletionException here. This should be part of the runAsync otherwise throw IOException or IllegalArgumentException.

Reply:
My vote would be throwing IllegalArgumentException within the runAsync

Reply:
Moved into runAsync and updated to throw IllegalArgumentException